### ISSUE TITLE
IPSec to and from a container outside the private network not working

### ISSUE THREAD

**Issue by rubentrancoso** [rancher/rancher#5347]
**Rancher Version:**
v1.1.0

**Docker Version:**
1.11.2

**OS and where are the hosts located? (cloud, bare metal, etc):**
CentOS 7 - 4 Hosts on private Lan - 1 Host on the cloud

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
single node rancher
lan hosts added with local resolution http://rancher.mydomain.lan
cloud host added with public resolution http://rancher.mydomain.com
nginx in front off rancher (both in the same DMZ machine)

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**

**on the from machine install rancher**

docker run -l io.rancher.container.network=true -d --restart=always --name=rancher-server -e "CATTLE_API_ALLOW_CLIENT_OVERRIDE=true" -e "CATTLE_HOST_API_PROXY_SCHEME=wss" -p 8080:8080 -v /mnt/nfs/rancher/lib/mysql:/var/lib/mysql -v /mnt/nfs/rancher/lib/cattle:/var/lib/cattle -v /mnt/nfs/rancher/log/mysql:/var/log/mysql  rancher/server

**install nginx**

docker run --restart=always -l io.rancher.container.network=true -d -p 80:80 -p 443:443 -v /mnt/nfs/nginx/certs:/etc/nginx/certs:ro -v /mnt/nfs/nginx/conf/nginx.conf:/etc/nginx/nginx.conf:ro -v /mnt/nfs/nginx/logs:/var/log/nginx --name=nginx nginx

set the rancher address to **https://rancher.mydomain.lan**
add hosts **lan** hosts

set the rancher address to **https://rancher.mydomain.com**
add **digtial ocean** host

**Results:** 
Everything but ipsec from Digital Ocean Host to private network works

**Expected:**
Establish a ipsec with all hosts, including the Digital Ocean one
Nginx and Rancher are running on the same host, and that host is exposed to the world through  DMZ with ports 500/4500 udp and 80/443 tcp open
All communication between containers on managed network works good. 
Communication from managed container outside the lan do not reach containers inside and vice versa.

**Besides that I found other issues maybe worth to be opened:**
1. Every time nginx configuration needs reload when for example a certificate gets updated, it get stuck and I got no response from port 80/443. Just made it work by stopping nginx container, deleting it, removing image, install again and restart firewalld
2. For some reason, rancher get opened as it has no password set. I did the password again and my environment was still there. (something related to the fact that the rancher volumes are mounted over nfs?)

NGINX.CONF

```
user  nginx;
worker_processes  1;

error_log  /var/log/nginx/error.log debug; # warn;
pid        /var/run/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       /etc/nginx/mime.types;
    default_type  application/octet-stream;

    log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  /var/log/nginx/access.log  main;

    sendfile        on;
    #tcp_nopush     on;

    keepalive_timeout  65;

    #gzip  on;

    include /etc/nginx/conf.d/*.conf;

   server {
            listen 80;
            server_name blog.mydomain.com jenkins.mydomain.com api.mydomain.com git.mydoemain.com;
            return 301 https://$server_name$request_uri;
   }
   server {
            listen 80;
            server_name *.mydomain.com mydomain.com;
            return 301 https://blog.mydomain.com;
   }

    server {
        listen 443 ssl http2;
        server_name blog.mydomain.com jenkins.mydomain.com api.mydomain.com;
        ssl_certificate  /etc/nginx/certs/fullchain.pem;
        ssl_certificate_key /etc/nginx/certs/privkey.pem;

        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_prefer_server_ciphers on;
        ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:E
CDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA2
56:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES
128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:E
CDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES
128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DS
S-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA25
6:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!E
XPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KR
B5-DES-CBC3-SHA';

        ssl_session_timeout 1d;
        ssl_session_cache shared:SSL:128m;
        ssl_stapling on;
        ssl_stapling_verify on;
        add_header Strict-Transport-Security "max-age=15768000; includeSubDomain
s";
        resolver 8.8.8.8;

        location / {
            proxy_pass            http://haproxy/;
            proxy_set_header      X-NginX-Proxy true;
            proxy_set_header      Host $http_host;
            proxy_set_header      X-Real-IP  $remote_addr;
            proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header      X-Forwarded-Proto https;

            proxy_redirect        default;
            proxy_read_timeout    15s;
            proxy_connect_timeout 15s;
        }
        location ~ /\. { deny  all; }

    }

    server {
        listen 443 ssl;
        server_name rancher.mydomain.com;
        ssl_certificate  /etc/nginx/certs/fullchain.pem;
        ssl_certificate_key /etc/nginx/certs/privkey.pem;

ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_prefer_server_ciphers on;
        ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:E
CDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA2
56:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES
128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:E
CDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES
128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DS
S-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA25
6:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!E
XPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KR
B5-DES-CBC3-SHA';

        ssl_session_timeout 1d;
        ssl_session_cache shared:SSL:128m;
        ssl_stapling on;
        ssl_stapling_verify on;
        add_header Strict-Transport-Security "max-age=15768000; includeSubDomain        

        resolver 8.8.8.8;

        location / {
            proxy_set_header Host $host;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_set_header X-Forwarded-Port $server_port;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_pass http://hosta.mydomain.lan:8080;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
    }

    upstream haproxy {
        server hosta.mydomain.lan:8081;
    }

}


#stream {
#
#    server {
#        listen 500 udp;
#        proxy_pass hb.mydomain.lan:500;
#        proxy_timeout 99999999s;
#        proxy_responses 0;
#    }
#
#    server {
#        listen 4500 udp;
#        proxy_pass hb.mydomain.lan:4500;
#        proxy_timeout 99999999s;
#        proxy_responses 0;
#    }
#
#}
```

**Comment by leodotcloud** [comment]
When you login to the Digital Ocean host, you should be able to reach the other hosts via their public IP addresses. How are the private hosts exposed to the public domain?

**Comment by rubentrancoso** [comment]
They are not. I have only a static IP for my domain and nginx in the front. 
What you said is new for me. It possibly clarify the behavior around IPSec with Rancher. 
But it's  not possible to do some kind of a NAT with udp since the external container knows the target IP and the front IP?

**Comment by vincent99** [comment]
The host agent connects to the rancher/server container(s) using the Host Registration URL.  All the hosts should be able to reach the server at the single configured address.  The server container is not involved in the overlay network/datapath between containers.

The overlay network is a [up-to] fully-connected mesh communicating directly between the involved hosts [as needed].  Network Agents know what Host the target container is on and reach that Host's Network Agent using the IP address the Host is registered as (which is shown on each in the UI).  There is no topology or routing discovery to derive a path by which host A can talk to host B indirectly via C -> D (or through the server container, or anything else).

So you have 2 island of hosts, which can talk to each other locally but not across the water to the other island.  You can put them in separate Environments so that there is no reason for them to try to talk to each other, or use scheduling labels in a single environment to assign services to one island or the other.

**Comment by rubentrancoso** [comment]
When I first tried this setup, all hosts were set using the same public URL, and what happened was that all lan hosts received the router internal IP on the rancher interface and IPSec did not work as well the reason I did the internal registration different from the other. 
Maybe a unencrypted VPN should do the trick. 
Would be nice to have something that could work as a router to the udp packages in those cases.
Looks like that should be a pretty common use case given that do not make much sense to have all hosts exposes to the world from what I see, and also not everyone can afford multiple IP addresses.
